// @hugoalh/advanced-determine/lib/main.mjs
import areEqual from "./are-equal.mjs";
import isArray from "./is-array.mjs";
import isBigInteger from "./is-big-integer.mjs";
import isFunction from "./is-function.mjs";
import isGenerator from "./is-generator.mjs";
import isJSON from "./is-json.mjs";
import isMap from "./is-map.mjs";
import isNumber from "./is-number.mjs";
import isObject from "./is-object.mjs";
import isPlainObject from "./is-plain-object.mjs";
import isRegularExpression from "./is-regular-expression.mjs";
import isSet from "./is-set.mjs";
import isString from "./is-string.mjs";
import isStringifyJSON from "./is-stringify-json.mjs";
import typeOf from "./type-of.mjs";
export {
	areEqual,
	isArray as isArr,
	isArray,
	isBigInteger as isBigInt,
	isBigInteger,
	isPlainObject as isDict,
	isPlainObject as isDictionary,
	isFunction as isFn,
	isFunction,
	isGenerator,
	isJSON,
	isStringifyJSON as isJSONStr,
	isStringifyJSON as isJSONStringified,
	isStringifyJSON as isJSONStringify,
	isArray as isList,
	isMap,
	isNumber as isNum,
	isNumber,
	isObject as isObj,
	isObject,
	isPlainObject as isObjectPlain,
	isPlainObject as isObjPlain,
	isPlainObject as isPlainObj,
	isPlainObject,
	isRegularExpression as isRegEx,
	isRegularExpression as isRegExp,
	isRegularExpression,
	isSet,
	isString as isStr,
	isString,
	isStringifyJSON as isStringifiedJSON,
	isStringifyJSON,
	isStringifyJSON as isStrJSON,
	typeOf
};
export default {
	areEqual,
	isArr: isArray,
	isArray,
	isBigInt: isBigInteger,
	isBigInteger,
	isDict: isPlainObject,
	isDictionary: isPlainObject,
	isFn: isFunction,
	isFunction,
	isGenerator,
	isJSON,
	isJSONStr: isStringifyJSON,
	isJSONStringified: isStringifyJSON,
	isJSONStringify: isStringifyJSON,
	isList: isArray,
	isMap,
	isNum: isNumber,
	isNumber,
	isObj: isObject,
	isObject,
	isObjectPlain: isPlainObject,
	isObjPlain: isPlainObject,
	isPlainObj: isPlainObject,
	isPlainObject,
	isRegEx: isRegularExpression,
	isRegExp: isRegularExpression,
	isRegularExpression,
	isSet,
	isStr: isString,
	isString,
	isStringifiedJSON: isStringifyJSON,
	isStringifyJSON,
	isStrJSON: isStringifyJSON,
	typeOf
};
